File: BlochMcConnellSplitter.py
import numpy as np

class BlochMcConnellSplitter:
    """Python translation of the MATLAB BlochMcConnellSplitter class.

    Usage:
        s = BlochMcConnellSplitter(offset, R1, R2, k, Meq, gamma, tau, b1x, b1y)
        M = s.integrate(M0)
    """

    def __init__(self, offset, R1, R2, k, Meq, gamma, tau, b1x, b1y):
        self.n_pools = int(np.asarray(offset).shape[0])
        self.tau = tau
        self.update_rotations(tau, b1x, b1y, gamma)
        self.update_relaxation_exchange(offset, R1, R2, k, Meq)

    def integrate(self, M0):
        """Integrate using asymmetric operator splitting.
        M0: 1D array of length 3*n_pools
        returns 1D array of length 3*n_pools
        """
        M = np.asarray(M0).reshape(1, -1).astype(np.complex128)
        num_steps = self.g_rot.shape[0]
        for i in range(num_steps):
            M = M @ self.g_rot[i]
            M = (M + self.RC) @ self.expRt - self.RC
        return np.real(M.T).ravel()

    def update_relaxation_exchange(self, offset, R1, R2, k, Meq):
        R = (self.relaxation_matrix(R1, R2) +
             self.exchange_matrix(k) +
             self.offset_matrix(offset))
        # Form C vector: zeros for transverse entries, R1 for longitudinal
        n = self.n_pools
        c = np.zeros(3 * n)
        for i in range(n):
            c[3 * i + 2] = R1[i]
        C = Meq * c
        # Solve R x = C
        x = np.linalg.solve(R, C)
        self.RC = x.reshape(1, -1)
        # Compute matrix exponential via diagonalization
        D, V = np.linalg.eig(R)
        expD = np.diag(np.exp(D * self.tau))
        expR = V @ expD @ np.linalg.inv(V)
        # Match MATLAB's transposed storage used in multiply M * expRt
        self.expRt = expR.T

    def update_rotations(self, tau, b1x, b1y, gamma):
        self.g_rot = self.create_rotations(self.n_pools, tau, np.asarray(b1x).ravel(), np.asarray(b1y).ravel(), gamma)

    @staticmethod
    def create_rotations(n_pools, tau, b1x, b1y, gamma):
        # b1x, b1y are arrays of length N (timesteps)
        b1x = np.asarray(b1x).ravel()
        b1y = np.asarray(b1y).ravel()
        N = b1x.size
        phi = -gamma * tau * np.abs(b1x - 1j * b1y)
        sin_phi = np.sin(phi)
        cos_phi = np.cos(phi)
        n1 = gamma * tau * b1x
        n2 = gamma * tau * b1y
        n3 = np.zeros_like(b1x)
        nonzero = np.abs(phi) > 0
        # avoid divide by zero
        denom = np.abs(phi)
        denom[~nonzero] = 1.0
        n1 = np.where(nonzero, n1 / denom, 0.0)
        n2 = np.where(nonzero, n2 / denom, 0.0)

        # Build 3x3 rotation matrices for each timestep
        rot3 = np.zeros((N, 3, 3), dtype=np.complex128)
        for i in range(N):
            n_vec = np.array([n1[i], n2[i], n3[i]])
            outer = np.outer(n_vec, n_vec)
            rot = (1 - cos_phi[i]) * outer + np.array([
                [cos_phi[i], -n3[i] * sin_phi[i], n2[i] * sin_phi[i]],
                [n3[i] * sin_phi[i], cos_phi[i], -n1[i] * sin_phi[i]],
                [-n2[i] * sin_phi[i], n1[i] * sin_phi[i], cos_phi[i]]
            ], dtype=np.complex128)
            # Transpose so block diagonal matches MATLAB permute(temp,[2,1,3]) behavior
            rot3[i] = rot.T

        # Create block-diagonal rotation for n_pools by Kron product
        big_rot = np.zeros((N, 3 * n_pools, 3 * n_pools), dtype=np.complex128)
        I = np.eye(n_pools)
        for i in range(N):
            big_rot[i] = np.kron(I, rot3[i])

        return big_rot

    @staticmethod
    def relaxation_matrix(R1, R2):
        n = int(np.asarray(R1).shape[0])
        n_dim = 3 * n
        R = np.zeros((n_dim, n_dim))
        for i in range(n):
            R[3 * i + 0, 3 * i + 0] = -R2[i]
            R[3 * i + 1, 3 * i + 1] = -R2[i]
            R[3 * i + 2, 3 * i + 2] = -R1[i]
        return R

    @staticmethod
    def offset_matrix(offset):
        # build block diagonal of 3x3 blocks
        blocks = []
        for off in np.asarray(offset).ravel():
            b = np.zeros((3, 3))
            b[1, 0] = off
            b[0, 1] = -off
            blocks.append(b)
        # block diagonal via kron
        n = len(blocks)
        M = np.zeros((3 * n, 3 * n))
        for i, b in enumerate(blocks):
            M[3 * i:3 * i + 3, 3 * i:3 * i + 3] = b
        return M

    @staticmethod
    def exchange_matrix(k):
        k = np.asarray(k)
        if k.shape[0] == 2:
            offdiag = np.block([
                [np.zeros((3, 3)), k[1, 0] * np.eye(3)],
                [k[0, 1] * np.eye(3), np.zeros((3, 3))]
            ])
            diagonal = np.block([
                [(-np.sum(k[0, :]) + k[0, 0]) * np.eye(3), np.zeros((3, 3))],
                [np.zeros((3, 3)), (-np.sum(k[1, :]) + k[1, 1]) * np.eye(3)]
            ])
        elif k.shape[0] == 3:
            offdiag = np.block([
                [np.zeros((3, 3)), k[1, 0] * np.eye(3), k[2, 0] * np.eye(3)],
                [k[0, 1] * np.eye(3), np.zeros((3, 3)), k[2, 1] * np.eye(3)],
                [k[0, 2] * np.eye(3), k[1, 2] * np.eye(3), np.zeros((3, 3))]
            ])
            diagonal = np.block([
                [(-np.sum(k[0, :]) + k[0, 0]) * np.eye(3), np.zeros((3, 3)), np.zeros((3, 3))],
                [np.zeros((3, 3)), (-np.sum(k[1, :]) + k[1, 1]) * np.eye(3), np.zeros((3, 3))],
                [np.zeros((3, 3)), np.zeros((3, 3)), (-np.sum(k[2, :]) + k[2, 2]) * np.eye(3)]
            ])
        else:
            raise ValueError("exchange_matrix only supports 2x2 or 3x3 k-matrices")
        return diagonal + offdiag

File: BlochMcConnellSplitter.m
classdef BlochMcConnellSplitter < handle
    % This class implements a matrix representation and solver for
    % Bloch-McConnell equations of a three-pool system. The class uses the
    % operator-splitting approach, which is outlined as follows:
    %
    % For each time interval in t, the relaxation/exchange part of the ODE is
    % solved exactly. The pulse part is solved with rotation matrices.
    % When the pulse does not change, the rotation matrices can be precomputed
    % and saved.
    % So the recipe is
    % 1. Compute rotation matrices for each timestep.
    % 2. Compute elements needed for the general solution of the
    %    relaxation/exchange part.
    %     2.1. Form the matrices Rxy and Rz and the vector Cz
    %     2.2. Solve for transverse relaxation/exchange
    %         2.2.1. Diagonalize Rxy
    %         2.2.2. Compute matrix exponential exp(Rxy*t)
    %     2.3. Solve for longitudinal relaxation/exchange 
    %         2.3.1. Diagonalize Rz
    %         2.3.2. Compute matrix exponential exp(Rz t)
    %         2.3.3. Compute Rz^-1*b
    %     2.4. Form relaxation/exchange matrix operator
    % 3. Loop through time and apply rotation and relaxation sequentially,
    %    according to the asymmetric operator splitting
    % 4. Return magnetization M
    % 
    % Here we assume that the timestep length is constant. This is not
    % necessary, but would require that the matrix exponentials are
    % computed for each timestep separately. They could still be
    % precomputed.
     

properties
    g_rot
    tau
    RC
    expRt
    n_pools
end


methods

    function obj = BlochMcConnellSplitter(offset, R1, R2, k, Meq, gamma, tau, b1x, b1y)
        obj.n_pools = size(offset, 1);
        obj.tau = tau;
        obj.update_rotations(tau, b1x, b1y, gamma);
        obj.update_relaxation_exchange(offset, R1, R2, k, Meq);
    end

    function M = integrate(obj, M0)
        % Integrates the Bloch-McConnell equation with asymmetric operator
        % splitting.
        M = M0.';
        % Loop through time
        for i=1:size(obj.g_rot, 3)
            M = M * obj.g_rot(:, :, i);
            M = (M + obj.RC) * obj.expRt  - obj.RC;
        end
        M = real(M.');
    end


    function update_relaxation_exchange(obj, offset, R1, R2, k, Meq)
        R = BlochMcConnellSplitter.relaxation_matrix(R1, R2) + ...
            BlochMcConnellSplitter.exchange_matrix(k) + ...
            BlochMcConnellSplitter.offset_matrix(offset);
        C = Meq .* reshape([zeros(2, obj.n_pools); R1'], obj.n_pools * 3, 1);
        obj.RC = (R\C).';
        [V, D] = eig(R);
        obj.expRt = (V * diag(exp(diag(D) * obj.tau)) * inv(V)).'; 
    end

    
    function update_rotations(obj, tau, b1x, b1y, gamma)
        obj.g_rot = obj.create_rotations(obj.n_pools, tau, b1x, b1y, gamma);
    end

end % methods

methods(Static)

    function rotation_matrices = create_rotations(n_pools, tau, b1x, b1y, gamma)
        % Assuming constant time interval between entries in B1.
        %% 1. compute phi and |phi|^-2
        % Should we use the endpoint B1 fields of each timestep? Probably
        %not.
        phi = -gamma * tau * abs(b1x - 1i*b1y);
        sin_phi = sin(phi);
        cos_phi = cos(phi);
        %% 2. compute n1, n2, n3

        n1 = gamma * tau * b1x;
        n2 = gamma * tau * b1y;
        n3 = zeros(length(b1x),1);
        n1(phi~=0) = n1(phi~=0) ./ abs(phi(phi~=0));
        n2(phi~=0) = n2(phi~=0) ./ abs(phi(phi~=0));
        %% 3. Assemble matrices
        rotation_matrices = [[n1.*n1, n1.*n2, n1.*n3];...
                             [n1.*n2, n2.*n2, n2.*n3];...
                             [n1.*n3, n2.*n3, n3.*n3]];
        rotation_matrices  = reshape(rotation_matrices, length(b1x), 3, 3);
        rotation_matrices = (1-cos_phi) .* rotation_matrices;  %% check that indexing is correct here.
        rotation_matrices = rotation_matrices + ...
            reshape([[cos_phi, -n3.*sin_phi, n2.*sin_phi];...
             [n3.*sin_phi, cos_phi, -n1.*sin_phi];...
             [-n2.*sin_phi, n1.*sin_phi, cos_phi]], length(b1x), 3, 3);
        rotation_matrices = permute(rotation_matrices, [2, 3, 1]);


        temp = zeros(n_pools*3, n_pools*3, size(rotation_matrices, 3));
        % Copy the same matrix to create a block diagonal matrix for
        % rotating all the pools.
        for i=0:n_pools-1
            temp(i*3 + 1: i*3 + 3, i*3 + 1: i*3 + 3, :) = rotation_matrices;
        end
        rotation_matrices = permute(temp, [2, 1, 3]);
    end


    function R = relaxation_matrix(R1, R2)
        n_dim = size(R1, 1)*3;
        R = zeros(n_dim, n_dim);
        for i=1:3:n_dim
            R(i, i) = -R2(ceil(i/3));
        end
        for i=2:3:n_dim
            R(i, i) = -R2(ceil(i/3));
        end
        for i=3:3:n_dim
            R(i, i) = -R1(ceil(i/3));
        end
    end


    function m_offset = offset_matrix(offset)
        for i = 1:size(offset, 1)
            blocks{i} = zeros(3);
            blocks{i}(2, 1) = offset(i);
            blocks{i}(1, 2) = -offset(i);
        end
        m_offset = blkdiag(blocks{:});
    end


    function m_k = exchange_matrix(k)
        switch size(k, 1)
            case 2
                offdiag = [zeros(3), k(2, 1).*eye(3);
                            k(1, 2).*eye(3), zeros(3)];
                diagonal = blkdiag((-sum(k(1, :)) + k(1,1)) .* eye(3),...
                                   (-sum(k(2, :)) + k(2,2)) .* eye(3));
            case 3
                offdiag = [zeros(3), k(2, 1).*eye(3), k(3, 1).*eye(3);
                            k(1, 2).*eye(3), zeros(3), k(3, 2).*eye(3);
                            k(1, 3).*eye(3), k(2, 3) .* eye(3), zeros(3)];
                diagonal = blkdiag((-sum(k(1, :)) + k(1,1)) .* eye(3),...
                                   (-sum(k(2, :)) + k(2,2)) .* eye(3),...
                                   (-sum(k(3, :)) + k(3,3)) .* eye(3));
        end
        m_k = diagonal + offdiag;
    end


end  % methods(Static)
end  % classdef